components/wifi/esp8266-driver/ESP8266Interface.cpp

```diff
@@ -568,15 +572,13 @@ int ESP8266Interface::socket_send(void *handle, const void *data, unsigned size)
         return NSAPI_ERROR_NO_SOCKET;
     }
 
-    unsigned long int sendStartTime = rtos::Kernel::get_ms_count();
-    do {
-        status = _esp.send(socket->id, data, size);
-    } while ((sendStartTime - rtos::Kernel::get_ms_count() < 50)
-            && (status != NSAPI_ERROR_OK));
+    status = _esp.send(socket->id, data, size);
 
-    if (status == NSAPI_ERROR_WOULD_BLOCK && socket->proto == NSAPI_TCP) {
-        tr_debug("ESP8266Interface::socket_send(): enqueuing the event call");
-        _global_event_queue->call_in(100, callback(this, &ESP8266Interface::event));
+    if (status == NSAPI_ERROR_WOULD_BLOCK
+        && socket->proto == NSAPI_TCP
+        && !core_util_atomic_flag_test_and_set(&_cbs[socket->id].deferred)) {
```

Comment:
Your needs are a bit more complex than atomic_flag can handle, sadly.

Believe this needs to be

    uint8_t cas_expected_current = false;

    && core_util_atomic_cas_u8(&_cbs[socket->id].deferred, &cas_expected_current, true)

Reply:
Fixed